<a href="https://colab.research.google.com/github/dotmanjohn/kosh/blob/master/Hyperparameter_Tuning_of_RF_Classifier_on_a_ISOLET_Dataset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Training a Random Forest Classifier with Hyperparameter Tuning on a ISOLET Dataset
You are working for a technology company and they are planning to launch a new voice assistant product. You have been tasked with building a classification model that will recognize the letters spelled out by a user based on the signal frequencies captured. Each sound can be captured and represented as a signal composed of multiple frequencies.

We will carry out hyperparameter tuning and select the model with the set of hyperparameters that gives the best performance (least overfitting model).

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

In [ ]:
file_url = 'https://raw.githubusercontent.com/PacktWorkshops/The-Data-Science-Workshop/master/Chapter04/Dataset/phpB0xrNj.csv'
df = pd.read_csv(file_url)
df.head()

f1      f2      f3      f4  ...    f615    f616    f617  class
0 -0.4394 -0.0930  0.1718  0.4620  ...  0.6410  0.5898 -0.4872    '1'
1 -0.4348 -0.1198  0.2474  0.4036  ...  0.4318  0.4546 -0.0910    '1'
2 -0.2330  0.2124  0.5014  0.5222  ...  0.2540  0.1588 -0.4762    '2'
3 -0.3808 -0.0096  0.2602  0.2554  ... -0.0072  0.0936 -0.1510    '2'
4 -0.3412  0.0946  0.6082  0.6216  ...  0.3124  0.2188 -0.2500    '3'

[5 rows x 618 columns]

In [ ]:
df.shape

(7797, 618)

In [ ]:
# Extract the class target variable into a new variable called y using the .pop() method:
y = df.pop('class')

Split the data into tranining ans testing sets

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(df, y, test_size=0.3, random_state=188)

Instantiate RandomForestClassifier with random_state=42 and then fit the model with the training set

In [ ]:
rf_model = RandomForestClassifier(random_state=42)
rf_model.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=42, verbose=0,
                       warm_start=False)

Predict the outcome of the training and testing sets with the .predict()method, save the results in a variable called 'train_preds' and 'test_preds'

In [ ]:
train_preds = rf_model.predict(X_train)
test_preds = rf_model.predict(X_test)

Calculate the accuracy score on the training and testing sets, save the result in a variable called train_acc and test_acc respectively and print their values

In [ ]:
train_acc = accuracy_score(y_train, train_preds)
test_acc = accuracy_score(y_test, test_preds)

In [ ]:
print(train_acc)
print(test_acc)

1.0
0.9320512820512821


Our model achieved a perfect accuracy score of 1 on the training set but 0.93 on the testing set. This means our model is overfitting and is not general enough. The ideal situation would be for the model to achieve a very similar, high-accuracy score on both the training and testing sets.

As a result of this, we will tune the hyperparameters of our model to get optimal values that will give us high-accuracy score that does not overfit.

In [ ]:
# Instantiate another RandomForestClassifier with random_state=42 and n_estimators=20, and then fit the model with the training set:
rf_model2 = RandomForestClassifier(random_state=42, n_estimators=20)
rf_model2.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=20,
                       n_jobs=None, oob_score=False, random_state=42, verbose=0,
                       warm_start=False)

Make predictions on the training and testing sets with .predict() and save the results into two new variables called train_preds2 and test_preds2

In [ ]:
train_preds2 = rf_model2.predict(X_train)
test_preds2 = rf_model2.predict(X_test)

# Calculate the accuracy score for the training and testing sets and save the results in two new variables called train_acc2 and test_acc2:
train_acc2 = accuracy_score(y_train, train_preds2)
test_acc2 = accuracy_score(y_test, test_preds2)

# Print the accuracy scores: train_acc and test_acc:
print(train_acc2)
print(test_acc2)

1.0
0.9226495726495727


The accuracy score decreased for the testing sets and now the difference is larger compared to the results from rf_model

Instantiate another RandomForestClassifier with random_state=42 and n_estimators=50, and then fit the model with the training set

In [ ]:
rf_model3 = RandomForestClassifier(random_state=42, n_estimators=50)
rf_model3.fit(X_train, y_train)

# Make predictions on the training and testing sets with .predict() and save the results into two new variables called train_preds3 and test_preds3:
train_preds3 = rf_model3.predict(X_train)
test_preds3 = rf_model3.predict(X_test)

# Calculate the accuracy score for the training and testing sets and save the results in two new variables called train_acc3 and test_acc3:
train_acc3 = accuracy_score(y_train, train_preds3)
test_acc3 = accuracy_score(y_test, test_preds3)

# Print the accuracy scores: train_acc3 and test_acc3:
print(train_acc3)
print(test_acc3)

1.0
0.926923076923077


This output shows us the model is still overfitting and the rf_model gives a better result compared to rf_model2 and rf_model3. Hence we will take our optimal value for n_estimator to be the default value i.e. 100

Now we will proceed to tuning max_depth

In [ ]:
# Instantiate RandomForestClassifier with random_state=42, n_estimators=100, max_depth=5 and then fit the model with the training set:
rf_model4 = RandomForestClassifier(random_state=42, n_estimators=100, max_depth=5)
rf_model4.fit(X_train, y_train)

# Make predictions on the training and testing sets with .predict() and save the results into two new variables called train_preds4 and test_preds4:
train_preds4 = rf_model4.predict(X_train)
test_preds4 = rf_model4.predict(X_test)
# Calculate the accuracy score for the training and testing sets and save the results in two new variables called train_acc4 and test_acc4:
train_acc4 = accuracy_score(y_train, train_preds4)
test_acc4 = accuracy_score(y_test, test_preds4)
# Print the accuracy scores: train_acc4 and test_acc4:
print(train_acc4)
print(test_acc4)

0.8704416345977644
0.8311965811965812


In [ ]:
# Instantiate RandomForestClassifier with random_state=42, n_estimators=50, max_depth=10 and then fit the model with the training set:
rf_model5 = RandomForestClassifier(random_state=42, n_estimators=100, max_depth=10)
rf_model5.fit(X_train, y_train)

train_preds5 = rf_model5.predict(X_train)
test_preds5 = rf_model5.predict(X_test)
train_acc5 = accuracy_score(y_train, train_preds5)
test_acc5 = accuracy_score(y_test, test_preds5)
print(train_acc5)
print(test_acc5)

0.9864394355873191
0.9260683760683761


max_depth=5 gives a better result hence, we will take this as our optimal value for max_depth

Now we will proceed to tuning min_leaf_samples

In [ ]:
# Instantiate RandomForestClassifier with random_state=42, n_estimators=30, max_depth=5, and min_samples_leaf=10, and then fit the model with the training set:
rf_model6 = RandomForestClassifier(random_state=42, n_estimators=100, max_depth=5, min_samples_leaf=10)
rf_model6.fit(X_train, y_train)

train_preds6 = rf_model6.predict(X_train)
test_preds6 = rf_model6.predict(X_test)
train_acc6 = accuracy_score(y_train, train_preds6)
test_acc6 = accuracy_score(y_test, test_preds6)
print(train_acc6)
print(test_acc6)

0.8697086311159978
0.8290598290598291


In [ ]:
# Instantiate RandomForestClassifier with random_state=42, n_estimators=30, max_depth=2, and min_samples_leaf=50, and then fit the model with the training set:
rf_model7 = RandomForestClassifier(random_state=42, n_estimators=100, max_depth=5, min_samples_leaf=50)
rf_model7.fit(X_train, y_train)

train_preds7 = rf_model7.predict(X_train)
test_preds7 = rf_model7.predict(X_test)
train_acc7 = accuracy_score(y_train, train_preds7)
test_acc7 = accuracy_score(y_test, test_preds7)
print(train_acc7)
print(test_acc7)

0.862561847168774
0.8282051282051283


min_samples_leaf=50 reduces the overfitting, hence this is our optimal value

Finally we will proceed to tuning max_features

In [ ]:
# Instantiate RandomForestClassifier with random_state=42, n_estimators=30, max_depth=2, and min_samples_leaf=50, max_features=0.5 and then fit the model with the training set:
rf_model8 = RandomForestClassifier(random_state=42, n_estimators=100, max_depth=5, min_samples_leaf=50, max_features=0.5)
rf_model8.fit(X_train, y_train)

train_preds8 = rf_model8.predict(X_train)
test_preds8 = rf_model8.predict(X_test)
train_acc8 = accuracy_score(y_train, train_preds8)
test_acc8 = accuracy_score(y_test, test_preds8)
print(train_acc8)
print(test_acc8)

0.7425325270295033
0.7175213675213675


In [ ]:
# Instantiate RandomForestClassifier with random_state=42, n_estimators=30, max_depth=2, and min_samples_leaf=50, max_features=0.3 and then fit the model with the training set:
rf_model9 = RandomForestClassifier(random_state=42, n_estimators=100, max_depth=5, min_samples_leaf=50, max_features=0.3)
rf_model9.fit(X_train, y_train)

train_preds9 = rf_model9.predict(X_train)
test_preds9 = rf_model9.predict(X_test)
train_acc9 = accuracy_score(y_train, train_preds9)
test_acc9 = accuracy_score(y_test, test_preds9)
print(train_acc9)
print(test_acc9)

0.7311709730621221
0.697008547008547


This final set of hyperparameters still doesn't achieve better results than the one we find with n_estimators=100, max_depth=5, min_samples_leaf=50, max_features=0.5.

We built several RandomForest classifier models that accurately predict the letters spoken from audio signals. We tried several values for the hyperparameters n_estimators, max_depth, min_samples_leaf, and max_features. The best combination of hyperparameters we came up with is n_estimators=100, max_depth=5, min_samples_leaf=50, and max_features='0.5'.

We achieved a final accuracy score of 0.74 for the training set and 0.71 for the testing set. The model is still slightly overfitting and could still be improved but it is a remarkable result.

**Our selected model is rf_model8**